src/pulsar_producers.erl

```diff
@@ -285,14 +276,24 @@ do_start_producer(#state{
                 {ok, URL} = pulsar_client:get_alive_pulsar_url(Pid),
                 {URL, BrokerServiceURL}
         end,
+    ParentPid = self(),
+    StateObserverCallback =
+        fun(ProducerState) ->
```

Comment:
we'll have to use `{fun ?MODULE:repport_state/2, [ParentPid]}` to make it hot-upgrade safe.